### ISSUE TITLE
Invalid UTF8: Couldn't convert Link tag bytes into String 

### ISSUE THREAD

**Issue by evomimic** [evomimic/map-holons#128]
### Reported Error:
In branch:  **_112-complete-load-core-schema-as-holons_**
While executing: **_test rstest_dance_tests::case_1_simple_add_related_holon_test_** 
I'm getting: 

```
2024-07-26T21:43:05.093623Z  INFO wasm_trace: dances::dancer:crates/coordinator/dances/src/dancer.rs:75 ======== RETURNING FROM  Dancer::dance() with DanceResponse {
    status_code: ServerError,
    description: MapString(
        "Invalid UTF8: Couldn't convert Link tag bytes[226, 130, 183, 65, 85, 84, 72, 79, 82, 69, 68, 95, 66, 89, 226, 138, 163, 0, 226, 147, 131, 107, 101, 121, 0, 226, 147, 11, 82, 111, 103, 101, 114, 66, 114, 105, 103, 103, 115, 0] into String (relationship name)",
    ),
    body: None,
    descriptor: None,
    staging_area: StagingArea {
        staged_holons: [],
        index: {},
    },
}
```

### Diagnosis:
The get_smartlink_from_link function within smartlink.rs is attempting to convert retrieved link_tag_bytes directly into a String using `from_utf8()`. But various separators used in our link_tags are not valid string characters.  

### Proposed Fix:
We need to decode the link_tag directly from its bytes (not attempt to first convert it into a String).

**Comment by evomimic** [comment]
Here's ChatGPT's recommendation for the decode function:

### Explanation:

1. Header Verification: The function starts by verifying the presence of the HEADER_BYTES.
2. Relationship Name Extraction: It extracts the relationship name by searching for the PROLOG_SEPERATOR.
3. Property Map Parsing: It iterates over the remaining bytes, extracting property names and values separated by the PROP_NAME_SEPERATOR, PROP_VAL_SEPERATOR, and UNICODE_NUL_STR.
4. Ensure all constants (HEADER_BYTES, PROLOG_SEPERATOR, etc.) match the encoding function. Adjust this implementation based on the actual structure of your encoded data.

### Rust Code
```
pub fn decode_link_tag(link_tag: LinkTag) -> (RelationshipName, Option<PropertyMap>) {
    let bytes = link_tag.0;
    let mut cursor = &bytes[..];

    if cursor.starts_with(&HEADER_BYTES) {
        cursor = &cursor[HEADER_BYTES.len()..];
    } else {
        panic!("Invalid LinkTag: Missing header bytes");
    }

    let name_end = cursor
        .windows(PROLOG_SEPERATOR.len())
        .position(|window| window == PROLOG_SEPERATOR)
        .unwrap();
    let relationship_name = str::from_utf8(&cursor[..name_end]).unwrap().to_string();
    cursor = &cursor[name_end + PROLOG_SEPERATOR.len()..];

    // Skip the first null character after the prolog separator
    cursor = &cursor[UNICODE_NUL_STR.len()..];

    let mut property_map = BTreeMap::new();

    while !cursor.is_empty() {
        if cursor.starts_with(&PROP_NAME_SEPERATOR) {
            cursor = &cursor[PROP_NAME_SEPERATOR.len()..];
        } else {
            break;
        }

        let prop_name_end = cursor
            .iter()
            .position(|&b| b == UNICODE_NUL_STR.as_bytes()[0])
            .unwrap();
        let prop_name = str::from_utf8(&cursor[..prop_name_end]).unwrap().to_string();
        cursor = &cursor[prop_name_end + UNICODE_NUL_STR.len()..];

        if cursor.starts_with(&PROP_VAL_SEPERATOR) {
            cursor = &cursor[PROP_VAL_SEPERATOR.len()..];
        } else {
            break;
        }

        let prop_val_end = cursor
            .iter()
            .position(|&b| b == UNICODE_NUL_STR.as_bytes()[0])
            .unwrap();
        let prop_val = cursor[..prop_val_end].to_vec();
        cursor = &cursor[prop_val_end + UNICODE_NUL_STR.len()..];

        property_map.insert(PropertyName((prop_name,)), PropertyValue(prop_val));
    }

    let relationship_name = RelationshipName((relationship_name,));
    let property_values = if property_map.is_empty() {
        None
    } else {
        Some(PropertyMap(property_map))
    };

    (relationship_name, property_values)
}

```

**Comment by evomimic** [comment]
Merged to main in PR #134

**Comment by evomimic** [comment]
Resolved
